File: backend/app/pds_client.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Protocol

import httpx

from app.config import settings


@dataclass(frozen=True)
class PDSDecisionRequest:
    actor_id: str
    action_type: str
    platform: str | None = None
    content: dict[str, Any] = field(default_factory=dict)
    context: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True)
class PDSDecision:
    decision_id: str
    verdict: str
    score: float = 0.0
    reasons: list[dict[str, Any]] = field(default_factory=list)
    evaluated_rules: list[str] = field(default_factory=list)
    rules_version: str = ""
    latency_ms: int = 0
    metadata: dict[str, Any] = field(default_factory=dict)


class PolicyDecisionClient(Protocol):
    async def decide(self, request: PDSDecisionRequest) -> PDSDecision:
        ...


class NoopPDSClient:
    async def decide(self, _request: PDSDecisionRequest) -> PDSDecision:
        return _fail_open("pds_disabled")


class PDSClient:
    def __init__(
        self,
        *,
        base_url: str | None = None,
        client_id: str | None = None,
        timeout_seconds: float | None = None,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        self.base_url = (settings.pds_base_url if base_url is None else base_url).rstrip("/")
        self.client_id = settings.pds_client_id if client_id is None else client_id
        self.timeout_seconds = settings.pds_timeout_seconds if timeout_seconds is None else timeout_seconds
        self._http_client = http_client

    async def decide(self, request: PDSDecisionRequest) -> PDSDecision:
        payload = {
            "actor_id": request.actor_id,
            "action": {"type": request.action_type, "platform": request.platform},
            "content": request.content,
            "context": request.context,
        }
        headers = {"X-Client-Id": self.client_id}
        url = f"{self.base_url}/v1/decide"

        try:
            response = await self._post(url, headers=headers, payload=payload)
        except httpx.RequestError:
            return _fail_open("pds_unavailable")

        if response.status_code >= 500:
            return _fail_open("pds_unavailable")

        response.raise_for_status()

        try:
            data = response.json()
            return _decision_from_payload(data)
        except (TypeError, ValueError):
            return _fail_open("pds_unavailable")

    async def _post(self, url: str, *, headers: dict[str, str], payload: dict[str, Any]) -> httpx.Response:
        if self._http_client is not None:
            return await self._http_client.post(
                url,
                json=payload,
                headers=headers,
                timeout=self.timeout_seconds,
            )

        async with httpx.AsyncClient(timeout=self.timeout_seconds) as client:
            return await client.post(url, json=payload, headers=headers)


def _decision_from_payload(data: Any) -> PDSDecision:
    if not isinstance(data, dict):
        raise TypeError("PDS decision response must be an object")

    return PDSDecision(
        decision_id=str(data.get("decision_id") or ""),
        verdict=str(data.get("verdict") or "allow"),
        score=float(data.get("score") or 0.0),
        reasons=_dict_list(data.get("reasons")),
        evaluated_rules=_str_list(data.get("evaluated_rules")),
        rules_version=str(data.get("rules_version") or ""),
        latency_ms=int(data.get("latency_ms") or 0),
        metadata=_dict(data.get("metadata")),
    )


def _dict(value: Any) -> dict[str, Any]:
    if value is None:
        return {}
    if not isinstance(value, dict):
        raise TypeError("expected object")
    return dict(value)


def _dict_list(value: Any) -> list[dict[str, Any]]:
    if value is None:
        return []
    if not isinstance(value, list) or not all(isinstance(item, dict) for item in value):
        raise TypeError("expected list of objects")
    return [dict(item) for item in value]


def _str_list(value: Any) -> list[str]:
    if value is None:
        return []
    if not isinstance(value, list):
        raise TypeError("expected list")
    return [str(item) for item in value]


def _fail_open(warning: str) -> PDSDecision:
    return PDSDecision(decision_id="", verdict="allow", metadata={"warning": warning})

File: internal/channelops/pds_client.go
package channelops

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"strings"
	"time"
)

const defaultPDSTimeout = 500 * time.Millisecond

var failPolicyByAction = map[string]string{
	"candidate_accept":    "allow",
	"plan_approval":       "flag",
	"publish":             "block",
	"promote_publication": "block",
}

type PDSDecisionRequest struct {
	ActorID    string         `json:"actor_id"`
	ActionType string         `json:"-"`
	Platform   string         `json:"-"`
	Content    map[string]any `json:"content"`
	Context    map[string]any `json:"context"`
}

type PDSDecision struct {
	DecisionID     string           `json:"decision_id"`
	Verdict        string           `json:"verdict"`
	Score          float64          `json:"score"`
	Reasons        []map[string]any `json:"reasons"`
	EvaluatedRules []string         `json:"evaluated_rules"`
	RulesVersion   string           `json:"rules_version"`
	LatencyMS      int              `json:"latency_ms"`
	Metadata       map[string]any   `json:"metadata"`
}

type PDSClient struct {
	Enabled     bool
	DevAllowAll bool
	BaseURL     string
	ClientID    string
	Timeout     time.Duration
	HTTPClient  *http.Client
}

func (c PDSClient) Decide(ctx context.Context, request PDSDecisionRequest) (PDSDecision, error) {
	if c.DevAllowAll {
		return PDSDecision{
			Verdict: "allow",
			Metadata: map[string]any{
				"warning":     "dev_allow_all",
				"fail_policy": "allow",
			},
		}, nil
	}

	if !c.Enabled {
		return failPolicyDecision(request.ActionType, "pds_disabled"), nil
	}

	payload := map[string]any{
		"actor_id": request.ActorID,
		"action": map[string]any{
			"type":     request.ActionType,
			"platform": request.Platform,
		},
		"content": mapOrEmpty(request.Content),
		"context": mapOrEmpty(request.Context),
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return failPolicyDecision(request.ActionType, "pds_parse_failed"), nil
	}

	httpRequest, err := http.NewRequestWithContext(ctx, http.MethodPost, strings.TrimRight(c.BaseURL, "/")+"/v1/decide", bytes.NewReader(body))
	if err != nil {
		return failPolicyDecision(request.ActionType, "pds_unavailable"), nil
	}
	httpRequest.Header.Set("Content-Type", "application/json")
	if strings.TrimSpace(c.ClientID) != "" {
		httpRequest.Header.Set("X-Client-Id", c.ClientID)
	}

	httpClient := c.HTTPClient
	if httpClient == nil {
		httpClient = &http.Client{Timeout: c.timeout()}
	}
	response, err := httpClient.Do(httpRequest)
	if err != nil {
		return failPolicyDecision(request.ActionType, "pds_unavailable"), nil
	}
	defer response.Body.Close()

	if response.StatusCode >= http.StatusBadRequest {
		_, _ = io.Copy(io.Discard, response.Body)
		return failPolicyDecision(request.ActionType, "pds_unavailable"), nil
	}

	var decision PDSDecision
	if err := json.NewDecoder(response.Body).Decode(&decision); err != nil {
		return failPolicyDecision(request.ActionType, "pds_parse_failed"), nil
	}
	verdict, ok := normalizeVerdict(decision.Verdict)
	if !ok {
		return failPolicyDecision(request.ActionType, "pds_parse_failed"), nil
	}
	decision.Verdict = verdict
	if decision.Metadata == nil {
		decision.Metadata = map[string]any{}
	}
	if decision.Reasons == nil {
		decision.Reasons = []map[string]any{}
	}
	if decision.EvaluatedRules == nil {
		decision.EvaluatedRules = []string{}
	}
	return decision, nil
}

func (c PDSClient) timeout() time.Duration {
	if c.Timeout > 0 {
		return c.Timeout
	}
	return defaultPDSTimeout
}

func failPolicyDecision(actionType string, warning string) PDSDecision {
	verdict := failPolicy(actionType)
	return PDSDecision{
		Verdict: verdict,
		Metadata: map[string]any{
			"warning":     warning,
			"fail_policy": verdict,
		},
	}
}

func failPolicy(actionType string) string {
	if verdict, ok := failPolicyByAction[actionType]; ok {
		return verdict
	}
	return "allow"
}

func normalizeVerdict(value string) (string, bool) {
	verdict := strings.ToLower(strings.TrimSpace(value))
	switch verdict {
	case "allow", "flag", "block":
		return verdict, true
	default:
		return "", false
	}
}

func mapOrEmpty(value map[string]any) map[string]any {
	if value == nil {
		return map[string]any{}
	}
	return value
}
